twoliter/embedded/imghelper

```diff
@@ -108,6 +110,228 @@ mkfs_data_xfs() {
+  rm "${shim}.unsigned"
+  pesign -i "${shim}" -l
+  pesigcheck -i "${shim}" -n 0 -c "${SBKEYS}/db.cer"
+
+  # Remove the original key if necessary, then sign the MOK manager as well.
+  if [[ "$(pesign -i "${mokm}" -l)" == "No signatures found." ]]; then
+    mv "${mokm}" "${mokm}.unsigned"
+  else
+    pesign -i "${mokm}" -o "${mokm}.unsigned" -u 0 -r
+  fi
+  pesign -i "${mokm}.unsigned" -o "${mokm}.signed" -s "${code_sign_key[@]}"
+  mv "${mokm}.signed" "${mokm}"
+  rm "${mokm}.unsigned"
+  pesign -i "${mokm}" -l
+  pesigcheck -i "${mokm}" -n 0 -c "${SBKEYS}/vendor.cer"
+
+  # Remove the original vendor key if necessary, replace the embedded gpg
+  # public key, then sign grub with the vendor key.
+  if [[ "$(pesign -i "${grub}" -l)" == "No signatures found." ]]; then
+    mv "${grub}" "${grub}.unsigned"
+  else
+    pesign -i "${grub}" -o "${grub}.unsigned" -u 0 -r
+  fi
+  "${arch}-objdumpcopy" "${grub}.unsigned" \
+    --file-alignment 4096 \
+    --update-section ".pubkey=${SBKEYS}/config-sign.pubkey"
+  pesign -i "${grub}.unsigned" -o "${grub}.signed" -s "${code_sign_key[@]}"
+  mv "${grub}.signed" "${grub}"
+  rm "${grub}.unsigned"
+  pesign -i "${grub}" -l
+  pesigcheck -i "${grub}" -n 0 -c "${SBKEYS}/vendor.cer"
+
+  popd >/dev/null || exit 1
+}
+
+uefi_dummysign() {
+  local efi_mount arch shims shim grubs grub
+  efi_mount="${1:?}"
+  arch="${2:?}"
+
+  pushd "${efi_mount}/EFI/BOOT" >/dev/null || exit 1
+
+  shims=(boot*.efi)
+  shim="${shims[0]}"
+  grubs=(grub*.efi)
+  grub="${grubs[0]}"
+
+  # Generate a zero-sized certificate in the expected format.
+  cert_table /dev/null "${SBKEYS}/vendor.obj"
+
+  # Replace the embedded vendor certificate with the zero-sized one, which
+  # shim will ignore when Secure Boot is disabled.
+  "${arch}-objdumpcopy" "${shim}" \
+    --update-section ".vendor_cert=${SBKEYS}/vendor.obj"
+
+  # Remove the embedded gpg public key to disable GRUB's signature checks.
+  "${arch}-objdumpcopy" "${grub}" \
+    --file-alignment 4096 \
+    --remove-section ".pubkey"
+
+  popd >/dev/null || exit 1
+}
+
+uefi_copycert() {
+  local efi_image
+  efi_image="${1:?}"
+  mcopy -i "${efi_image}" "${SBKEYS}"/db.{crt,cer} ::/EFI/BOOT
+}
+
+vmlinuz_sbsetup() {
+  local vmlinuz
+  local -n code_sign_key
+  vmlinuz="${1:?}"
+  code_sign_key="${2:?}"
+  if [[ "$(pesign -i "${vmlinuz}" -l)" == "No signatures found." ]]; then
+    mv "${vmlinuz}" "${vmlinuz}.unsigned"
+  else
+    pesign -i "${vmlinuz}" -o "${vmlinuz}.unsigned" -u 0 -r
+  fi
+  pesign -i "${vmlinuz}.unsigned" -o "${vmlinuz}.signed" -s "${code_sign_key[@]}"
+  mv "${vmlinuz}.signed" "${vmlinuz}"
+  rm "${vmlinuz}.unsigned"
+  pesign -i "${vmlinuz}" -l
+  pesigcheck -i "${vmlinuz}" -n 0 -c "${SBKEYS}/vendor.cer"
+}
+
+vmlinuz_hmac() {
+  local vmlinuz
+  vmlinuz="${1:?}"
+  openssl sha512 -hmac FIPS-FTW-RHT2009 -hex "${vmlinuz}" |
+    awk -v vmlinuz="${vmlinuz}" '{ print $2 "  " vmlinuz }' \
+      >"${vmlinuz%/*}.${vmlinuz##*/}.hmac"
+}
+
+sign_grub() {
```

Comment:
The function is named `sign_grub` but you're not actually signing grub but rather `grub.cfg` ...